pkg/egressgateway/manager.go

```diff
@@ -543,141 +538,96 @@ func (manager *Manager) updatePoliciesMatchedEndpointIDs() {
-			egressPolicies[*key] = *val
-		})
-
-	addEgressRule := func(endpointIP netip.Addr, dstCIDR netip.Prefix, excludedCIDR bool, gwc *gatewayConfig) {
-		policyKey := egressmap.NewEgressPolicyKey4(endpointIP, dstCIDR)
-		policyVal, policyPresent := egressPolicies[policyKey]
+	// To ensure no connectivity disruption, first add/update rules,
+	// then delete obsolete ones.
 
-		gatewayIP := gwc.gatewayIP
-		if excludedCIDR {
-			gatewayIP = ExcludedCIDRIPv4
-		}
+	// Pull in the current contents of the bpf map.
+	bpfEgressState := map[egressmap.EgressPolicyKey4]egressmap.EgressPolicyVal4{}
+	manager.policyMap.IterateWithCallback(func(key *egressmap.EgressPolicyKey4, val *egressmap.EgressPolicyVal4) {
+		bpfEgressState[*key] = *val
+	})
 
-		if policyPresent && policyVal.Match(gwc.egressIP, gatewayIP) {
-			return
+	// Add and update from desired.
+	for key, val := range desired {
+		var (
+			srcIP       = key.GetSourceIP()
+			dstCIDR     = key.GetDestCIDR()
+			egressAddr  = val.GetEgressAddr()
+			gatewayAddr = val.GetGatewayAddr()
+		)
+
+		currentVal, exists := bpfEgressState[key]
+		if exists && currentVal.Match(egressAddr, gatewayAddr) {
+			continue
 		}
 
 		logger := log.WithFields(logrus.Fields{
-			logfields.SourceIP:        endpointIP,
-			logfields.DestinationCIDR: dstCIDR.String(),
-			logfields.EgressIP:        gwc.egressIP,
-			logfields.GatewayIP:       gatewayIP,
+			logfields.SourceIP:        srcIP,
+			logfields.DestinationCIDR: dstCIDR,
+			logfields.EgressIP:        egressAddr,
+			logfields.GatewayIP:       gatewayAddr,
 		})
 
-		if err := manager.policyMap.Update(endpointIP, dstCIDR, gwc.egressIP, gatewayIP); err != nil {
+		if err := manager.policyMap.Update(srcIP, dstCIDR, egressAddr, gatewayAddr); err != nil {
 			logger.WithError(err).Error("Error applying egress gateway policy")
 		} else {
 			logger.Debug("Egress gateway policy applied")
 		}
 	}
 
-	for _, policyConfig := range manager.policyConfigs {
-		policyConfig.forEachEndpointAndCIDR(addEgressRule)
-	}
-}
-
-// removeUnusedEgressRules is responsible for removing any entry in the egress policy BPF map which
-// is not baked by an actual k8s CiliumEgressGatewayPolicy.
-func (manager *Manager) removeUnusedEgressRules() {
-	egressPolicies := map[egressmap.EgressPolicyKey4]egressmap.EgressPolicyVal4{}
-	manager.policyMap.IterateWithCallback(
-		func(key *egressmap.EgressPolicyKey4, val *egressmap.EgressPolicyVal4) {
-			egressPolicies[*key] = *val
-		})
-
-nextPolicyKey:
-	for policyKey, policyVal := range egressPolicies {
-		matchPolicy := func(endpointIP netip.Addr, dstCIDR netip.Prefix, excludedCIDR bool, gwc *gatewayConfig) bool {
-			gatewayIP := gwc.gatewayIP
-			if excludedCIDR {
-				gatewayIP = ExcludedCIDRIPv4
-			}
-
-			return policyKey.Match(endpointIP, dstCIDR) && policyVal.Match(gwc.egressIP, gatewayIP)
+	// Remove stuff that's not present in the desired state.
+	for key, val := range bpfEgressState {
+		if _, exists := desired[key]; exists {
```

Comment:
I think the original logic in `removeUnusedEgressRules` checks if both policy key and value matches the desired state. Is it OK not to do the same here?

Reply:
I believe so. The code slightly further up from here will Add and also update existing keys, so all keys that should exist in the datapath will be present by this point in the code. (and also contain the correct value)

The highlighted part of the code simply makes sure that any keys that are present in the datapath that aren't desired, get deleted. For deletion, we don't care what value that held, we already know that the key must be deleted.